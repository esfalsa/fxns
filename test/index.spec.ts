import { beforeAll, afterEach, describe, it, expect } from 'vitest';
import { fetchMock } from 'cloudflare:test';
import * as fixtures from './fixtures';
import app from '../src/index';

beforeAll(() => {
	fetchMock.activate();
	fetchMock.disableNetConnect();
});
afterEach(() => fetchMock.assertNoPendingInterceptors());

const BASE = 'https://fxns.pronoun.workers.dev/';

function appFetch(
	path: string,
	opts?: RequestInit<IncomingRequestCfProperties<unknown>>,
) {
	return app.fetch(
		new Request(new URL(path, BASE), opts),
		undefined,
		undefined as unknown as ExecutionContext,
	) as Promise<Response>;
}

function ogTag(property: string, content: string) {
	return `<meta property="og:${property}" content="${content}" />`;
}

describe('healthcheck (/)', async () => {
	const res = await appFetch('/');

	it('responds with redirect status 302', () => {
		expect(res.status).toBe(302);
	});

	it('redirects to the GitHub repository', () => {
		expect(res.headers.get('Location')).toBe('https://github.com/esfalsa/fxns');
	});
});

describe('favicon request', async () => {
	for (const path of [
		'/favicon.ico',
		'/favicon-32x32.png',
		'/favicon-16x16.png',
	]) {
		const res = await appFetch(path);

		it('responds with redirect status 301', () => {
			expect(res.status).toBe(301);
		});

		it('redirects to the NationStates favicon', () => {
			expect(res.headers.get('Location')).toBe(
				`https://www.nationstates.net${path}`,
			);
		});
	}
});

describe('apple touch icon request', async () => {
	const res = await appFetch('/apple-touch-icon.png');

	it('responds with redirect status 301', () => {
		expect(res.status).toBe(301);
	});

	it('redirects to the NationStates apple touch icon', () => {
		expect(res.headers.get('Location')).toBe(
			'https://www.nationstates.net/apple-touch-icon.png',
		);
	});
});

describe('android chrome icon request', async () => {
	const res = await appFetch('/android-chrome-192x192.png');

	it('responds with redirect status 301', () => {
		expect(res.status).toBe(301);
	});

	it('redirects to the NationStates apple touch icon', () => {
		expect(res.headers.get('Location')).toBe(
			'https://www.nationstates.net/android-chrome-192x192.png',
		);
	});
});

describe('nation opengraph request', async () => {
	fetchMock
		.get('https://www.nationstates.net')
		.intercept({ path: /.*/ })
		.reply(200, fixtures.nation);

	const res = await appFetch('/nation=esfalsa', {
		headers: { 'User-Agent': 'curl/8.4.0' },
	});
	const body = await res.text();

	it('lists the nation as the title', () => {
		expect(body).toContain(ogTag('title', 'The Flagless Nation of Esfalsa'));
	});

	it("lists the nation's flag as the image", () => {
		expect(body).toContain(
			ogTag(
				'image',
				'https://www.nationstates.net/images/flags/uploads/esfalsa__372439.svg',
			),
		);
	});
});

describe('region opengraph request', async () => {
	fetchMock
		.get('https://www.nationstates.net')
		.intercept({ path: /.*/ })
		.reply(200, fixtures.region);

	const res = await appFetch('/region=the_south_pacific', {
		headers: { 'User-Agent': 'curl/8.4.0' },
	});
	const body = await res.text();

	it('matches snapshopt', () => {
		expect(body).toMatchSnapshot();
	});
});

describe('proposal opengraph request', async () => {
	fetchMock
		.get('https://www.nationstates.net')
		.intercept({ path: /.*/ })
		.reply(200, fixtures.proposals);

	const res = await appFetch(
		`/page=UN_view_proposal/id=${fixtures.standardProposalID}`,
		{
			headers: { 'User-Agent': 'curl/8.4.0' },
		},
	);
	const body = await res.text();

	it('matches snapshopt', () => {
		expect(body).toMatchSnapshot();
	});
});

describe('nation 404 request', async () => {
	fetchMock
		.get('https://www.nationstates.net')
		.intercept({ path: /.*/ })
		.reply(404, fixtures.nation404);

	const res = await appFetch('/nation=thisnationdoesnotexist', {
		headers: { 'User-Agent': 'curl/8.4.0' },
	});
	const body = await res.text();

	it('responds with status 404', () => {
		expect(res.status).toBe(404);
	});

	it('matches snapshopt', () => {
		expect(body).toMatchSnapshot();
	});
});
