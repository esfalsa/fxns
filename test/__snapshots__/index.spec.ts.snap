// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nation 404 request > matches snapshopt 1`] = `"{"status":404,"error":"Not Found"}"`;

exports[`proposal opengraph request > matches snapshopt 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="UTF-8" />
<link rel="canonical" href="https://www.nationstates.net/page=UN_view_proposal/id=westinor_1718510253" />
<title>Proposal | Commend Nasicournia</title>
<meta name="description" content="A Commendation proposal by Westinor created on June 15, 2024 with 69 approvals.

Legal: 2 | Illegal: 0 | Discard: 0" />
<!-- required Open Graph metadata -->
<meta property="og:title" content="Proposal | Commend Nasicournia" />
<meta property="og:type" content="website" />
<meta property="og:image" content="https://www.nationstates.net/images/waflag.svg" />
<meta property="og:url" content="https://www.nationstates.net/page=UN_view_proposal/id=westinor_1718510253" />
<!-- optional Open Graph metadata -->
<meta property="og:description" content="A Commendation proposal by Westinor created on June 15, 2024 with 69 approvals.

Legal: 2 | Illegal: 0 | Discard: 0" />
<meta property="og:site_name" content="NationStates" />
<!-- structured properties Open Graph metadata -->
<meta property="og:image:alt" content="World Assembly logo" />
<!-- non-standard Open Graph metadata -->
<meta property="og:ignore_canonical" content="true" />
<!-- Twitter metadata -->
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:site" content="@NationStates" />
<meta name="twitter:description" content="A Commendation proposal by Westinor created on June 15, 2024 with 69 approvals.

Legal: 2 | Illegal: 0 | Discard: 0" />
<meta name="twitter:title" content="Proposal | Commend Nasicournia" />
<meta name="twitter:image" content="https://www.nationstates.net/images/waflag.svg" />
<meta name="twitter:image:alt" content="World Assembly logo" />
</head>
<body></body>
</html>"
`;

exports[`region opengraph request > matches snapshopt 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="UTF-8" />
<link rel="canonical" href="https://www.nationstates.net/region=the_south_pacific" />
<title>the South Pacific</title>
<meta name="description" content="A gargantuan feeder with 4,822 nations and extremely high regional power, led by Delegate Ebonhand." />
<!-- required Open Graph metadata -->
<meta property="og:title" content="the South Pacific" />
<meta property="og:type" content="website" />
<meta property="og:image" content="https://www.nationstates.net/images/flags/uploads/rflags/the_south_pacific__652104.png" />
<meta property="og:url" content="https://www.nationstates.net/region=the_south_pacific" />
<!-- optional Open Graph metadata -->
<meta property="og:description" content="A gargantuan feeder with 4,822 nations and extremely high regional power, led by Delegate Ebonhand." />
<meta property="og:site_name" content="NationStates" />
<!-- structured properties Open Graph metadata -->
<meta property="og:image:alt" content="Flag of the South Pacific" />
<!-- non-standard Open Graph metadata -->
<meta property="og:ignore_canonical" content="true" />
<!-- Twitter metadata -->
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:site" content="@NationStates" />
<meta name="twitter:description" content="A gargantuan feeder with 4,822 nations and extremely high regional power, led by Delegate Ebonhand." />
<meta name="twitter:title" content="the South Pacific" />
<meta name="twitter:image" content="https://www.nationstates.net/images/flags/uploads/rflags/the_south_pacific__652104.png" />
<meta name="twitter:image:alt" content="Flag of the South Pacific" />
</head>
<body></body>
</html>"
`;
